`invalid-name` type alias false-positive for union values

### Bug description

With something like:

```python
# pylint: disable=missing-docstring
from typing import Union
x: Union[str, int] = 42
```

pylint claims a wrong type alias name, with no type aliases involved.

### Configuration

_No response_

### Command used

```shell
pylint test.py
```


### Pylint output

```shell
************* Module test
.../test.py:3:0: C0103: Type alias name "x" doesn't conform to predefined naming style (invalid-name)
```


### Expected behavior

No output

### Pylint version

```shell
pylint 2.17.0
astroid 2.15.0
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201]

also reproducible with the current main branches:

pylint 3.0.0a6
astroid 2.16.0dev0
Python 3.10.10 (main, Mar  5 2023, 22:26:53) [GCC 12.2.1 20230201]
```


### OS / Environment

_No response_

### Additional dependencies

_No response_

--- a/doc/whatsnew/fragments/8487.false_positive
+++ b/doc/whatsnew/fragments/8487.false_positive
@@ -1,0 +1,3 @@
+No longer consider ``Union`` as type annotation as type alias for naming checks.
+
+Closes #8487

--- a/pylint/checkers/base/name_checker/checker.py
+++ b/pylint/checkers/base/name_checker/checker.py
@@ -600,12 +600,10 @@ class NameChecker(_BasicChecker):
                 # Union is a special case because it can be used as a type alias
                 # or as a type annotation. We only want to check the former.
                 assert node is not None
-                if (
+                return not (
                     isinstance(node.parent, nodes.AnnAssign)
                     and node.parent.value is not None
-                ):
-                    return False
-                return True
+                )
         elif isinstance(inferred, nodes.FunctionDef):
             if inferred.qname() == "typing.TypeAlias":
                 return True

--- a/pylint/checkers/base/name_checker/checker.py
+++ b/pylint/checkers/base/name_checker/checker.py
@@ -597,6 +597,14 @@ class NameChecker(_BasicChecker):
         inferred = utils.safe_infer(node)
         if isinstance(inferred, nodes.ClassDef):
             if inferred.qname() == ".Union":
+                # Union is a special case because it can be used as a type alias
+                # or as a type annotation. We only want to check the former.
+                assert node is not None
+                if (
+                    isinstance(node.parent, nodes.AnnAssign)
+                    and node.parent.value is not None
+                ):
+                    return False
                 return True
         elif isinstance(inferred, nodes.FunctionDef):
             if inferred.qname() == "typing.TypeAlias":

--- a/tests/functional/t/typealias_naming_style_default.py
+++ b/tests/functional/t/typealias_naming_style_default.py
@@ -21,3 +21,7 @@ BAD_NAME = Union[int, str]  # [invalid-name]
 _BAD_NAME = Union[int, str]  # [invalid-name]
 __BAD_NAME = Union[int, str]  # [invalid-name]
 ANOTHERBADNAME = Union[int, str]  # [invalid-name]
+
+# Regression tests
+# This is not a TypeAlias, and thus shouldn't flag the message
+x: Union[str, int] = 42

